src/Razor/src/Microsoft.AspNetCore.Razor.Language/RazorHtmlLoweringPass.cs

```diff
@@ -0,0 +1,280 @@
+            }
+
+            public override void VisitMarkupTagHelperAttribute(MarkupTagHelperAttributeSyntax node)
+            {
+                WriteNode(node, isHtml: true, base.VisitMarkupTagHelperAttribute);
+            }
+
+            public override void VisitMarkupTagHelperAttributeValue(MarkupTagHelperAttributeValueSyntax node)
+            {
+                WriteNode(node, isHtml: true, base.VisitMarkupTagHelperAttributeValue);
+            }
+
+            public override void VisitMarkupTagHelperDirectiveAttribute(MarkupTagHelperDirectiveAttributeSyntax node)
+            {
+                WriteNode(node, isHtml: true, base.VisitMarkupTagHelperDirectiveAttribute);
+            }
+
+            public override void VisitMarkupTagHelperElement(MarkupTagHelperElementSyntax node)
+            {
+                WriteNode(node, isHtml: true, base.VisitMarkupTagHelperElement);
+            }
+
+            public override void VisitMarkupTagHelperEndTag(MarkupTagHelperEndTagSyntax node)
+            {
+                WriteNode(node, isHtml: true, base.VisitMarkupTagHelperEndTag);
+            }
+
+            public override void VisitMarkupTagHelperStartTag(MarkupTagHelperStartTagSyntax node)
+            {
+                WriteNode(node, isHtml: true, base.VisitMarkupTagHelperStartTag);
+            }
+
+            public override void VisitMarkupTextLiteral(MarkupTextLiteralSyntax node)
+            {
+                WriteNode(node, isHtml: true, base.VisitMarkupTextLiteral);
+            }
+
+            public override void VisitMarkupTransition(MarkupTransitionSyntax node)
+            {
+                WriteNode(node, isHtml: false, base.VisitMarkupTransition);
+            }
+
+            public override void VisitRazorCommentBlock(RazorCommentBlockSyntax node)
+            {
+                WriteNode(node, isHtml: false, base.VisitRazorCommentBlock);
+            }
+
+            public override void VisitRazorDirective(RazorDirectiveSyntax node)
+            {
+                WriteNode(node, isHtml: false, base.VisitRazorDirective);
+            }
+
+            public override void VisitRazorDocument(RazorDocumentSyntax node)
+            {
+                WriteNode(node, isHtml: true, base.VisitRazorDocument);
+            }
+
+            public override void VisitRazorMetaCode(RazorMetaCodeSyntax node)
+            {
+                WriteNode(node, isHtml: false, base.VisitRazorMetaCode);
+            }
+
+            public override void VisitUnclassifiedTextLiteral(UnclassifiedTextLiteralSyntax node)
+            {
+                WriteNode(node, isHtml: false, base.VisitUnclassifiedTextLiteral);
+            }
+
+            public override void VisitToken(SyntaxToken token)
+            {
+                base.VisitToken(token);
+                WriteToken(token);
+            }
+
+            private void WriteToken(SyntaxToken token)
```

Comment:
I'm taking the simple approach of replacing non-HTML with spaces. I know this could have unexpected effects on HTML formatting but we can tackle that when we get there.